lib/AST/ProtocolConformance.cpp

```diff
@@ -1059,3 +1059,151 @@ DeclContext::getLocalConformances(
 
   return result;
 }
+
+/// Check of all types used by the conformance are canonical.
+bool ProtocolConformance::isCanonical() const {
+  // Normal conformances are considered canonical by
+  // construction.
+  if (getKind() != ProtocolConformanceKind::Normal) {
+    if (!getType()->isCanonical())
+      return false;
+    if (!getInterfaceType()->isCanonical())
+      return false;
+  }
+
+  switch (getKind()) {
+  case ProtocolConformanceKind::Normal: {
+    auto normalConformance = cast<NormalProtocolConformance>(this);
+    auto sigConformances = normalConformance->getSignatureConformances();
+    for (auto conf : sigConformances)
+      if (!conf.isCanonical())
+        return false;
+    return true;
+  }
+  case ProtocolConformanceKind::Inherited: {
+    // Substitute the base.
+    auto inheritedConformance
+      = cast<InheritedProtocolConformance>(this);
+    return inheritedConformance->getInheritedConformance()->isCanonical();
+  }
+  case ProtocolConformanceKind::Specialized: {
+    // Substitute the substitutions in the specialized conformance.
+    auto spec = cast<SpecializedProtocolConformance>(this);
+    auto genericConformance = spec->getGenericConformance();
+    if (!genericConformance->isCanonical())
+      return false;
+    auto specSubs = spec->getGenericSubstitutions();
+    for (auto sub : specSubs) {
+      if (!sub.getReplacement()->isCanonical())
```

Comment:
Maybe there should be a `Substitution::isCanonical()` instead of open coding it here?